src/layer.rs

```diff
@@ -1,102 +1,732 @@
+    pub render: js_sys::Function,
 }
 
-impl From<&str> for LayoutProperty {
-    fn from(value: &str) -> Self {
-        LayoutProperty::String(value.into())
+impl IntoLayer for CustomLayer {
+    fn into_layer(self) -> Layer {
+        Layer::Custom(self)
     }
 }
 
-impl From<String> for LayoutProperty {
-    fn from(value: String) -> Self {
-        LayoutProperty::String(value)
+fn make_wasm_closure(some_fn: impl Fn(JsValue, JsValue) + 'static) -> js_sys::Function {
+    Closure::<dyn Fn(JsValue, JsValue)>::new(some_fn)
+        .into_js_value()
+        .into()
+}
+
+impl CustomLayer {
+    pub fn new<F>(id: impl Into<String>, source: impl Into<String>, render: F) -> CustomLayer
+    where
+        F: 'static + Fn(JsValue, JsValue),
+    {
+        let render = Closure::<dyn Fn(JsValue, JsValue)>::new(render);
+        CustomLayer {
+            inner: LayerBase::new(id, source),
+            rendering_mode: None,
+            // Use no-ops for the optional functions
+            on_add: make_wasm_closure(|_1, _2| {}),
+            on_remove: make_wasm_closure(|_1, _2| {}),
+            prerender: make_wasm_closure(|_1, _2| {}),
+            render: render.into_js_value().into(),
+        }
+    }
+    pub fn set_on_add(&mut self, on_add_fn: impl Fn(JsValue, JsValue) + 'static) {
+        self.on_add = make_wasm_closure(on_add_fn);
+    }
+    pub fn set_on_remove(&mut self, on_remove_fn: impl Fn(JsValue, JsValue) + 'static) {
+        self.on_remove = make_wasm_closure(on_remove_fn);
+    }
+    pub fn set_render(&mut self, render_fn: impl Fn(JsValue, JsValue) + 'static) {
+        self.render = make_wasm_closure(render_fn);
+    }
+    pub fn set_prerender(&mut self, prerender_fn: impl Fn(JsValue, JsValue) + 'static) {
+        self.prerender = make_wasm_closure(prerender_fn);
     }
 }
 
-impl From<f64> for LayoutProperty {
-    fn from(value: f64) -> Self {
-        LayoutProperty::Number(value)
+impl<T: fmt::Debug> fmt::Display for Expression<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Expression::String(s) => write!(f, "\"{}\"", s),
+            Expression::Number(n) => write!(f, "{}", n),
+            Expression::Bool(b) => write!(f, "{}", b),
+            Expression::List(list) => {
+                write!(f, "[")?;
+                for (index, item) in list.iter().enumerate() {
+                    if index > 0 {
+                        write!(f, ", ")?;
+                    }
+                    write!(f, "{}", item)?;
+                }
+                write!(f, "]")
+            }
+            Expression::Enum(e) => write!(f, "{:?}", e),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FillLayer {
+    #[serde(flatten)]
+    pub inner: LayerBase,
+    #[serde(default)]
+    pub layout: FillLayout,
+    #[serde(default)]
+    pub paint: FillPaint,
+}
```

Comment:
This is _the_ typical layer. It has a layer base with stuff like, source, max and min zoom etc. all flattened directly into the layer. And it has strongly typed layout and paint properties that default to basically all `None`'s quite readable I think and very usable in combination with autocomplete!

The only downside is that setting the source for example is `my_layer.inner.source = …` instead of `my_layer.source = …` what do you think?

Reply:
Yeah I actually agree. I will change that.